Mathlib/ModelTheory/LanguageMap.lean

```diff
@@ -500,6 +529,51 @@ instance map_constants_inclusion_isExpansionOn :
 
 end WithConstants
 
+variable (L L')
+
+/-- If `L.Expands L'`, then there is a privileged `LHom` from `L'` into `L`. -/
+class Expands where
+  /-- The privileged map `L' →ᴸ L`, preferably accessed as `L'.inclusion L`. -/
+  toLHom : L' →ᴸ L
+  toLHom_injective : toLHom.Injective
+
+/-- Note that the variables of `L.inclusion L'` are reversed from the corresponding instance
+  `L'.Expands L`. -/
+protected abbrev inclusion [L'.Expands L] : L →ᴸ L' := Expands.toLHom
+
+lemma inclusion_injective [L'.Expands L] : (L.inclusion L').Injective := Expands.toLHom_injective
+
+namespace Expands
+
+variable (L'' : Language)
+
+/-- If `L.Expands L'` and `L'.Expands L''`, then `L.Expands L''`, by composing the default
+  inclusions. -/
+def trans [L.Expands L'] [L'.Expands L''] : L.Expands L'' where
+  toLHom := (L'.inclusion L).comp (L''.inclusion L')
+  toLHom_injective := (L'.inclusion_injective L).comp (L''.inclusion_injective L')
+
+variable {L L' L''} [L.Expands L']
+
+@[simps]
+instance : L.Expands L where
+  toLHom := (LEquiv.refl L).toLHom
+  toLHom_injective := (LEquiv.refl L).toHom_injective
+
+instance : (L.sum L'').Expands L' where
+  toLHom := LHom.sumInl.comp (L'.inclusion L)
+  toLHom_injective := LHom.sumInl_injective.comp (L'.inclusion_injective L)
+
+instance : (L''.sum L).Expands L' where
+  toLHom := LHom.sumInr.comp (L'.inclusion L)
+  toLHom_injective := LHom.sumInr_injective.comp (L'.inclusion_injective L)
```

Comment:
Is this not a recipe for diamonds? Should `(L.sum L).Expands L` be the inclusion into the left component, or the right one?

Reply:
It is. I don't think this will ever actually come up with the kinds of languages that we will actually implement in mathlib, but there are few enough of those that these can both be made `def`s anyway.